code/segmentedtickerplant/stplog.q

```diff
@@ -0,0 +1,250 @@
+
+// Sequence number
+seqnum:0
+
+// Functions to add columns on updates
+updtab:@[value;`.stplg.updtab;enlist[`]!enlist {(enlist(count first x)#y),x}]
+
+// If set to memorybatch, publish and write to disk will be run in batches
+// insert to table in memory, on a timer flush the table to disk and publish, update counts
+upd[`memorybatch]:{[t;x;now]
+  t insert updtab[t] . (x;now);
+ };
+
+zts[`memorybatch]:{
+  {[t]
+    if[count value t;
+      `..loghandles[t] enlist (`upd;t;value flip value t);
+      @[`.stplg.msgcount;t;+;1];
+      @[`.stplg.rowcount;t;+;count value t];
+      .stpps.pubclear[t]];
+  }each .stpps.t;
+ };
+
+// Standard batch mode - write to disk immediately, publish in batches
+upd[`defaultbatch]:{[t;x;now]
+  t insert x:.stplg.updtab[t] . (x;now);
+  `..loghandles[t] enlist(`upd;t;x);
+  // track tmp counts, and add these after publish
+  @[`.stplg.tmpmsgcount;t;+;1];
+  @[`.stplg.tmprowcount;t;+;count first x];
+ };
+
+zts[`defaultbatch]:{
+  // publish and clear all tables, increment counts
+  .stpps.pubclear[.stpps.t];
+  // after data has been published, updated the counts
+  .stplg.msgcount+:.stplg.tmpmsgcount;
+  .stplg.rowcount+:.stplg.tmprowcount;
+  // reset temp counts
+  .stplg.tmpmsgcount:.stplg.tmprowcount:()!();
+ };
+
+// Immediate mode - publish and write immediately
+upd[`immediate]:{[t;x;now]
+  x:updtab[t] . (x;now);
+  `..loghandles[t] enlist(`upd;t;x);
+  @[`.stplg.msgcount;t;+;1];
+  @[`.stplg.rowcount;t;+;count first x];
+  .stpps.pub[t;x]
+ };
+
+zts[`immediate]:{}
+
+//////////////////////////////////////////////////////////////////////////////////////
+
+// Functions to obtain logs for client replay ////////////////////////////////////////
+// replaylog called from client-side, returns nested list of logcounts and lognames
+replaylog:{[t]
+  getlogs[replayperiod][t]
+ }
+
+getlogs:enlist[`]!enlist ()
+
+// If replayperiod set to `period, only replay logs for current logging period
+getlogs[`period]:{[t]
+  distinct flip (.stplg.msgcount;exec tbl!logname from `..currlog where tbl in t)@\:t
+ };
+
+// If replayperiod set to `day, replay all of today's logs
+getlogs[`day]:{[t]
+  lnames:distinct uj/[{
+    select seq,tbls,logname,msgcount from .stpm.metatable where x in/: tbls
+    }each t];
+  // Meta table does not store counts for live logs, so these are populated here
+  lnames:update msgcount:sum each .stplg.msgcount[tbls] from lnames where seq=.stplg.i;
+  flip value exec `long$msgcount,logname from lnames
+ };
+
+//////////////////////////////////////////////////////////////////////////////////////
+
+// Open log for a single table at start of logging period
+openlog:{[multilog;dir;tab;p]
+  lname:logname[multilog][dir;tab;p];
+  .lg.o[`openlog;"opening logfile: ",string lname];
+  h:$[(notexists:not type key lname)or null h0:exec first handle from `..currlog where logname=lname;
+    [.[if[notexists;lname;();:;()]];hopen lname];
```

Comment:
Is this logic correct? I find this a bit confusing to read but this doesn't seem right: `if[notexists;lname;();:;()]`